src/main/java/uk/ac/cam/cl/dtg/isaac/quiz/PgQuestionAttempts.java

```diff
@@ -231,46 +224,91 @@ public Map<String, Map<String, List<QuestionValidationResponse>>> getQuestionAtt
+                String inParams = userIds.stream().map(u -> "?").collect(Collectors.joining(", "));
+                query.append(String.format(" user_id IN (%s)", inParams));
+            }
+            query.append(" ORDER BY \"timestamp\" ASC");
+
+            try (PreparedStatement pst = conn.prepareStatement(query.toString())) {
+
+                Map<Long, Map<String, Map<String, List<LightweightQuestionValidationResponse>>>> mapToReturn = Maps.newHashMap();
+
+                int index = 1;
+                for (Long userId : userIds) {
+                    pst.setLong(index++, userId);
+                    mapToReturn.put(userId, new HashMap<>());
+                }
+
+                try (ResultSet results = pst.executeQuery()) {
+                    while (results.next()) {
+                        LightweightQuestionValidationResponse partialQuestionAttempt = resultsToLightweightValidationResponse(results);
+
+                        String questionPageId = extractPageIdFromQuestionId(partialQuestionAttempt.getQuestionId());
+                        String questionId = partialQuestionAttempt.getQuestionId();
+                        Long userId = results.getLong("user_id");
+
+                        Map<String, Map<String, List<LightweightQuestionValidationResponse>>> mapOfQuestionAttemptsByPage
+                                = mapToReturn.get(userId);
+
+                        Map<String, List<LightweightQuestionValidationResponse>> attemptsForThisQuestionPage =
+                                mapOfQuestionAttemptsByPage.computeIfAbsent(questionPageId, k -> Maps.newHashMap());
+
+                        List<LightweightQuestionValidationResponse> listOfResponses =
+                                attemptsForThisQuestionPage.computeIfAbsent(questionId, k -> Lists.newArrayList());
+
+                        listOfResponses.add(partialQuestionAttempt);
+                    }
+                    return mapToReturn;
+                }
+            }
+        } catch (SQLException e) {
+            throw new SegueDatabaseException("Postgres exception", e);
+        }
+    }
     
     @Override
     public Map<Long, Map<String, Map<String, List<LightweightQuestionValidationResponse>>>>
-            getQuestionAttemptsByUsersAndQuestionPrefix(final List<Long> userIds, final List<String> questionPageIds)
+            getQuestionAttemptsByUsersAndQuestionPrefix(final List<Long> userIds, final List<String> allQuestionPageIds)
             throws SegueDatabaseException {
-        Validate.notEmpty(questionPageIds);
-        
+        Validate.notEmpty(allQuestionPageIds);
         if (userIds.isEmpty()) {
             return Maps.newHashMap();
         }
 
+        List<String> uniquePageIds = allQuestionPageIds.stream().distinct().collect(Collectors.toList());
+        if (uniquePageIds.size() > MAX_PAGE_IDS_TO_MATCH) {
+            // The repeated "OR question_id LIKE ___" will get very inefficient beyond this:
```

Comment:
This query will always use the existing `question_attempts_by_user_question` index (that's `btree (user_id, question_id text_pattern_ops)`) for a bitmap scan (it can't be index-only scan because all the columns we need aren't in the index).

Then it runs through the final results comparing each row to the OR clauses, and this gets slow if there are more than ~100-200 OR clauses. Since we're going to load them into a HashMap in Java and then access them by ID, it turned out to be faster to just leave the filtering to the HashMap if there really are lots of question IDs to filter by.
(I.e. we can either run through `n` rows in Postgres and compare it to a list of `m` question IDs - `O(n*m)`, or else we can not filter in Postgres and add each row to a HashMap - `O(n)` and then do the HashMap accesses we'd do either way).

Reply:
I do wonder what would happen if we had an inverted index `btree (question_id text_pattern_ops, user_id)` and if that'd help. But our queries are:
 - ~10 questions for many users (single assignment progress); this query excels.
 - many questions for one user (My Gameboards / My Assignments); this query excels for low-activity users, and we fall back to just "all attempts" for highly active users (but using Lightweight responses in this PR, not the full ones we currently do).
 - many questions for many users (group progress); there's no good way to do this, but filtering just by user seems the most reasonable since the `question x user` space is sparse.

Reply:
That's more or less what I was saying, minus the index then. Might as well not do any limiting in the database and do the filtering in Java.